File: python/lexer.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-


from pava_exception import *
import re


class Lexer(object):
    """
    Lexical analyzer for Pava
    """
    _REGEX = r'\s*((//.*)|([0-9]+)|("(\\"|\\\\|\\n|[^"])*")|[A-Z_a-z][A-Z_a-z0-9]*' \
             '|==|<=|>=|&&|\|\||[!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~])?'

    def __init__(self, lines):
        self._lines = lines
        self._pattern = re.compile(Lexer._REGEX)
        self._token_queue = []
        self._current_line = 0

    @staticmethod
    def _convert_string(string):
        s = string[1:len(string) - 1]
        return s.replace(r'\\', '\\').replace(r'\n', '\n').replace(r'\"', '\"')

    def read(self):
        if self._fill_queue(0):
            return self._token_queue.pop(0)
        else:
            return EOFToken()

    def peak(self, position):
        if self._fill_queue(position):
            return self._token_queue[position]
        else:
            return EOFToken()

    def _fill_queue(self, position):
        while position >= len(self._token_queue):
            if self._current_line < len(self._lines):
                self._read_line()
            else:
                return False

        return True

    def _read_line(self):
        line = self._lines[self._current_line]
        self._current_line += 1

        start_pos = 0
        end_pos = len(line)
        while start_pos < end_pos:
            match = self._pattern.match(line, start_pos, end_pos)
            if match is not None:
                self._add_token(self._current_line, match)
                start_pos = match.end()
            else:
                raise ParseException("bad token at line " + str(self._current_line))

        self._token_queue.append(IdentifierToken(self._current_line, Token.EOL))

    def _add_token(self, line_number, match):
        if match.group(1) is not None and match.group(2) is None:
            if match.group(3) is not None:
                token = NumberToken(line_number, int(match.group(3)))
            elif match.group(4) is not None:
                token = StringToken(line_number, self._convert_string(match.group(4)))
            else:
                token = IdentifierToken(line_number, match.group(1))
            self._token_queue.append(token)

File: src/Pava/Lexer.java
package Pava;

import Pava.Token.IdToken;
import Pava.Token.NumToken;
import Pava.Token.StrToken;
import Pava.Token.Token;

import java.io.IOException;
import java.io.LineNumberReader;
import java.io.Reader;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Lexical analyzer for Pava
 * Created by yangruihan on 2016/11/20.
 */
public class Lexer {

    public static String regexPat = "\\s*((//.*)|([0-9]+)|(\"(\\\\\"|\\\\\\\\|\\\\n|[^\"]) *\")" +
            "|[A-Za-z][A-Za-z0-9]*|==|<=|>=|&&|\\|\\||\\p{Punct})?";

    private Pattern pattern = Pattern.compile(regexPat);
    private ArrayList<Token> queue = new ArrayList<>();
    private boolean hasMore;
    private LineNumberReader reader;

    public Lexer(Reader r) {
        hasMore = true;
        reader = new LineNumberReader(r);
    }

    public Token read() throws ParseException {
        if (fillQueue(0)) {
            return queue.remove(0);
        } else {
            return Token.EOF;
        }
    }

    public Token peek(int i) throws ParseException {
        if (fillQueue(i)) {
            return queue.get(i);
        } else {
            return Token.EOF;
        }
    }

    private boolean fillQueue(int i) throws ParseException {
        while (i >= queue.size()) {
            if (hasMore) {
                readLine();
            } else {
                return false;
            }
        }
        return true;
    }

    protected void addToken(int lineNo, Matcher matcher) {
        String m = matcher.group(1);
        if (m != null) {        // if not a space
            if (matcher.group(2) == null) {     // if not a comment
                Token token;
                if (matcher.group(3) != null) {
                    token = new NumToken(lineNo, Integer.parseInt(m));
                } else if (matcher.group(4) != null) {
                    token = new StrToken(lineNo, toStringLiteral(m));
                } else {
                    token = new IdToken(lineNo, m);
                }
                queue.add(token);
            }
        }
    }

    /**
     * converts an escape character to a normal character
     * "\\" -> '\'
     * "\n" -> '\n'
     * "\"" -> '"'
     *
     * @param s origin string
     * @return converted string
     */
    protected String toStringLiteral(String s) {
        StringBuilder sb = new StringBuilder();
        int len = s.length() - 1;
        for (int i = 0; i < len; i++) {
            char c = s.charAt(i);
            if (c == '\\' && i + 1 < len) {
                char c2 = s.charAt(i + 1);
                if (c2 == '\\' || c2 == '"') {
                    c = s.charAt(++i);
                } else if (c2 == 'n') {
                    ++i;
                    c = '\n';
                }
            }
            sb.append(c);
        }

        return sb.toString();
    }

    private void readLine() throws ParseException {
        String line;
        try {
            line = reader.readLine();
        } catch (IOException e) {
            throw new ParseException(e);
        }
        if (line == null) {
            hasMore = false;
            return;
        }

        int lineNo = reader.getLineNumber();
        Matcher matcher = pattern.matcher(line);
        matcher.useTransparentBounds(true).useAnchoringBounds(false);
        int pos = 0;
        int endPos = line.length();
        while (pos < endPos) {
            matcher.region(pos, endPos);
            if (matcher.lookingAt()) {
                addToken(lineNo, matcher);
                pos = matcher.end();
            } else {
                throw new ParseException("bad token at line " + lineNo);
            }
        }

        queue.add(new IdToken(lineNo, Token.EOL));
    }
}
